### ISSUE TITLE
Smoothing backtracks uneccessarily

### ISSUE THREAD

**Issue by hjabbot** [antarctica/PolarRoute#291]
I was testing splitting conditions for a vector field and came across this example where the smoothing is very obviously bugged. This was found on the `kml_route_output` branch. ~~Dijkstra paths failed to be created on `main` so I can't test on that, but since it appears to be a smoothing bug, I imagine the issue persists there too.~~ _**EDIT:** Don't know what happened before, but I can run it on main now, and the issue doesn't exist there_

The bug I'm highlighting is the fact that the smoothing backtracks on the non-uniform mesh in the below plot. The smoothed route also looks remarkably pointy in the uniform mesh given how smooth the input vector field is, so I imagine that's also an issue.

Green path is Dijkstra, red path is smoothed. Waypoints are from bottom left to top right.
Colourmap on righthand plot is the curl of the vector field, with the actual vector field shown by the arrows. From this we can see that there is no reason for the route to backtrack on itself.
![Visualisation](https://github.com/user-attachments/assets/6d4cbbd9-3341-4cb1-b73e-14c57278230f)

Here are the configs to recreate this case.
[failing_case.zip](https://github.com/user-attachments/files/16727263/failing_case.zip)

**Comment by Ulvetanna** [comment]
Ok perfect, put this on my Tuesday list to look into 😄

**Comment by Ulvetanna** [comment]
Looking into the smoothing bug at comparing to tag `0.5` the path converges without an issue creating. 

Further discussion of this issue needs to go into making sure that the smoothed pathed regression tests don't introduce smoothing bugs in `0.6`. Running the regression tests there seems to be no issue between the two outputs (see attached file). The easiest way to check for differences would be outputting the dijkstra graphs that go into the path smoothing, I will looking into this next


![Screenshot 2024-08-27 at 15 13 01](https://github.com/user-attachments/assets/a325df65-b651-4fa7-b147-da633a3e26b0)
![Screenshot 2024-08-27 at 15 12 23](https://github.com/user-attachments/assets/fdb17af6-6cd6-4bd7-8c27-6b1ce10444cf)
[path_smoothing_regressiontest.txt](https://github.com/user-attachments/files/16763673/path_smoothing_regressiontest.txt)

**Comment by hjabbot** [comment]
Potentially related? Found a failing case late in the day
[another_failing_case.zip](https://github.com/user-attachments/files/16765533/another_failing_case.zip)

**Comment by gecoombs** [comment]
Did you learn any more by comparing the Dijkstra graphs @Ulvetanna? We should probably sit down at some point this week and look at the function that generates the `initialised_dijkstra_graph` on v0.6. As you said, it's possible that this bug originates there and I was hoping to find ways to improve its speed as well.

I assume the pre-smoothing Dijkstra paths match on both versions for this case?

**Comment by gecoombs** [comment]
This bug seems to have been caused by the way keys were generated for the dictionary that stores the crossing points and travel time legs calculated during Dijkstra for use in the smoothing. In some cases it was possible for these to be non-unique and some values to be overwritten. I've changed how these are named on the linked branch to avoid this possibility and when I run this example I get identical results to `main`.
